packages/syft/src/syft/service/code_history/code_history_service.py

```diff
@@ -0,0 +1,261 @@
+from ..code.user_code import SubmitUserCode
+from ..code.user_code import UserCode
+from ..context import AuthedServiceContext
+from ..response import SyftError
+from ..response import SyftSuccess
+from ..service import AbstractService
+from ..service import service_method
+from ..user.user_roles import DATA_OWNER_ROLE_LEVEL
+from ..user.user_roles import DATA_SCIENTIST_ROLE_LEVEL
+from .code_history import CodeHistory
+from .code_history import CodeHistoryDict
+from .code_history import CodeVersions
+from .code_history import UserHistoryDict
+from .code_history_stash import CodeHistoryStash
+
+
+@instrument
+@serializable()
+class CodeHistoryService(AbstractService):
+    store: DocumentStore
+    stash: CodeHistoryStash
+
+    def __init__(self, store: DocumentStore) -> None:
+        self.store = store
+        self.stash = CodeHistoryStash(store=store)
+
+    @service_method(
+        path="code_history.submit_version",
+        name="submit_version",
+        roles=DATA_SCIENTIST_ROLE_LEVEL,
+    )
+    def submit_version(
+        self,
+        context: AuthedServiceContext,
+        code: Union[SubmitUserCode, UserCode],
+        comment: Optional[str] = None,
+        add_request: Optional[bool] = True,
+        reason: Optional[str] = None,
+    ) -> Union[SyftSuccess, SyftError]:
+        user_code_service = context.node.get_service("usercodeservice")
+
+        final_result = None
+        if isinstance(code, SubmitUserCode):
+            if add_request:
+                result = user_code_service.request_code_execution(
+                    context=context, code=code, reason=reason
+                )
+                if isinstance(result, SyftError):
+                    return result
+                code = (
+                    result.ok()
+                    .changes[0]
+                    .linked_obj.resolve_with_context(context=context)
+                    .ok()
+                )
+                final_result = result
+            else:
+                result = user_code_service.submit(context=context, code=code)
+                if isinstance(result, SyftError):
+                    return result
+                uid = UID.from_string(result.message.split(" ")[-1])
+                code = user_code_service.get_by_uid(context=context, uid=uid)
+
+        elif isinstance(code, UserCode):
+            result = user_code_service.get_by_uid(context=context, uid=code.id)
+            if isinstance(result, SyftError):
+                return result
+            code = result
+
+        result_code_history_list = self.stash.get_by_service_func_name(
+            credentials=context.credentials, service_func_name=code.service_func_name
+        )
+
+        if result_code_history_list.is_err():
+            return SyftError(message=result_code_history_list.err())
+
+        code_history = None
+        code_history_list = result_code_history_list.ok()
+
+        for elem in code_history_list:
```

Comment:
this logic should be handled by the stash, ideally it would also just use both keys (service_func_name and user verify key) to search instead of loop

Reply:
maybe similar to how we do it in get_by_func_name_and_user_email?

Reply:
yep, it was an oversight on our part